### PR TITLE
push images with multi-arch to support amd64/arm64

### LINKED ISSUE #181: ARM versions available ?
Do you know if ARM images exist ?
I try that but it fails : 
```
$ docker pull gcr.io/heptio-images/sonobuoy-arm:v0.10.0
Error response from daemon: manifest for gcr.io/heptio-images/sonobuoy-arm:v0.10.0 not found
```

### PR DESCRIPTION
**Pull request by lubinsz**
Signed-off-by: Bin Lu <bin.lu@arm.com>

**What this PR does / why we need it**:
add arm64 support 

**Which issue(s) this PR fixes**
- Fixes #181

**Special notes for your reviewer**:

**Release note**:
```
release-note
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -91,6 +127,31 @@ push:
 		$(DOCKER) push $(REGISTRY)/$(TARGET):latest; \
 	fi
 
+push_manifest:
+	/go/bin/manifest-tool push from-args --platforms $(PLATFORMS) --template $(REGISTRY)/$(TARGET)-ARCH:$(VERSION) --target  $(REGISTRY)/$(TARGET):$(VERSION)
+
+push: pre container
+	for arch in $(LINUX_ARCH); do \
+		$(MAKE) push_images TARGET="sonobuoy-$$arch"; \
```
Comment (liztio) [inline Makefile]: `An image does not exist locally with the tag: gcr.io/heptio-images/sonobuoy-amd64`

When should these be created?

File: Dockerfile [hunk 2]
```diff
@@ -12,11 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM golang:1.10-alpine
+FROM BASEIMAGE
```
Comment (liztio) [inline Dockerfile]: I think seperate Dockerfiles would be easier to reason about

File: Makefile [hunk 3]
```diff
@@ -71,17 +76,48 @@ lint:
...
+       -t $(REGISTRY)/$(TARGET):$(IMAGE_BRANCH) \
+       -t $(REGISTRY)/$(TARGET):$(GIT_REF) \
+       -f $(DOCKERFILE) \
 		.
 
-sonobuoy:
-	$(DOCKER_BUILD) '$(BUILD)'
+container: sonobuoy
+	for arch in $(LINUX_ARCH); do \
+		if [ $$arch = amd64 ]; then \
```
[the remark this answers is not in the document]
Comment (lubinsz) [reply Makefile]: hi @liztio 
In my opinion, inconvenient maintenance in the later stage if use seperate Dockerfiles.
For example: if we need to add a new feature in Dockerfile, we should modify them twice.
3 differences between the two substitutions:
1, baseimage
2, command: In x86, we can't use any 'RUN' command in Dockerfile to build target for CROSS-platform(arm, power). 
3, binary file: different binary files correspond to different platforms

File: Dockerfile (hunk 2 above)
Comment (lubinsz) [reply Dockerfile]: Inconvenient maintenance in the later stage if use seperate Dockerfiles as I think.
Also, same approach in k8s project: single Dockerfile, and use 'sed' to change it for each platform.

File: Makefile [hunk 4]
```diff
@@ -71,17 +76,48 @@ lint:
...
+			sed 's|BASEIMAGE|alpine:3.7|g' Dockerfile > Dockerfile-$$arch; \
```
[the remark this answers is not in the document]
Comment (lubinsz) [reply Makefile]: Hi, 
I have fixed it. Please check it.

File: Makefile [hunk 5]
```diff
@@ -91,6 +127,31 @@ push:
 		$(DOCKER) push $(REGISTRY)/$(TARGET):latest; \
 	fi
 
+push_manifest:
+	/go/bin/manifest-tool push from-args --platforms $(PLATFORMS) --template $(REGISTRY)/$(TARGET)-ARCH:$(VERSION) --target  $(REGISTRY)/$(TARGET):$(VERSION)
```
[the remark this answers is not in the document]
Comment (lubinsz) [reply Makefile]: Hi,
I have fixed it. Please check it.

File: Makefile [hunk 6]
```diff
@@ -12,26 +12,33 @@
...
+LINUX_ARCH := amd64 arm64
+DOCKERFILE :=
+PLATFORMS := $(subst $(SPACE),$(COMMA),$(foreach arch,$(LINUX_ARCH),linux/$(arch)))
 
 GIT_VERSION ?= $(shell git describe --always --dirty)
 IMAGE_VERSION ?= $(shell git describe --always --dirty)
 IMAGE_BRANCH ?= $(shell git rev-parse --abbrev-ref HEAD | sed 's/\///g')
 GIT_REF = $(shell git rev-parse --short=8 --verify HEAD)
 
+GOBIN = /go/bin
```
Comment (liztio) [inline Makefile]: I don't want to need to be root to run this script. `GOBIN ?= /go/bin` maybe? Are you building this on a machine that doesn't have go set up?

Comment (lubinsz) [reply Makefile]: hi Liztio,
I have updated it, please check it.
Thanks.

**Comment by lubinsz** [comment]
@liztio 
Any other question?

File: Makefile [hunk 7]
```diff
@@ -71,17 +81,49 @@ lint:
 vet:
 	$(DOCKER_BUILD) '$(VET)'
 
-container: sonobuoy
+pre:
+	curl -sSL https://github.com/estesp/manifest-tool/releases/download/$(MANIFEST_TOOL_VERSION)/manifest-tool-linux-amd64 > $(MANIFEST_TOOL_DIR)/manifest-tool
```
Comment (liztio) [inline Makefile]: This doesn't work on OSX: `/var/folders/v9/30xb9zqj20nf56xbvmz6qpym0000gn/T/tmp.cdp3AjCw/manifest-tool: /var/folders/v9/30xb9zqj20nf56xbvmz6qpym0000gn/T/tmp.cdp3AjCw/manifest-tool: cannot execute binary file`

Why did you switch from `go get`?

Comment (lubinsz) [reply Makefile]: @liztio 
"go get" is back. Please check it.

**Comment by liztio** [comment]
Hi there @lubinsz! It took me a little bit of fiddling this morning but I managed to get a test push to work. Because I work from osx, I needed to make the following change:

```diff
diff --git a/Makefile b/Makefile
index 40a982d..965c831 100644
--- a/Makefile
+++ b/Makefile
@@ -129,7 +129,7 @@ push_images:
        fi

 push_manifest:
-       $(GOPATH)/bin/manifest-tool push from-args --platforms $(PLATFORMS) --template $(REGISTRY)/$(TARGET)-ARCH:$(VERSION) --target  $(REGISTRY)/$(TARGET):$(VERSION)
+       $(GOPATH)/bin/manifest-tool -username oauth2accesstoken --password "`gcloud auth print-access-token`" push from-args --platforms $(PLATFORMS) --template $(REGISTRY)/$(TARGET)-ARCH:$(VERSION) --target  $(REGISTRY)/$(TARGET):$(VERSION)

 push: pre container
        for arch in $(LINUX_ARCH); do \
```

If you could add that, we should be good to add this!

**Comment by lubinsz** [comment]
@liztio 
Done.

**Comment by liztio** [comment]
Sorry for all the back and forth! We should be cutting v1.11.5 soon, and hopefully that'll be the first release that supports arm!

**Comment by kalyxin02** [comment]
@liztio Could we know when will the v1.11.5 release happen? Thanks!

**Comment by liztio** [comment]
@kalyxin02 we recently release v1.11.6  due to an issue with .5, it should be live now, give it a go!

**Comment by johnSchnake** [comment]
@lubinsz I'm just looking into the makefile/images a bit since we need to update them for #609 

Why did we need to use different base images at all? From my short trial, it seems like we can just use the basic `GOOS=<> GOARCH=<> go build` and the binary created from my mac was even able to run on an `arm64v8:busybox` container.

I'm just wondering if I'll only notice the need in a certain situation; I dont actually have an arm64 device to actually test on either.

**Pull request merged**